PersistentStore/PersistentStore.cpp

```diff
@@ -0,0 +1,709 @@
+
+            namespaces.clear();
+
+            if (db)
+            {
+                sqlite3_stmt *stmt;
+                sqlite3_prepare_v2(db, "SELECT name FROM namespace;", -1, &stmt, NULL);
+
+                while (sqlite3_step(stmt) == SQLITE_ROW)
+                    namespaces.push_back((const char*)sqlite3_column_text(stmt, 0));
+
+                sqlite3_finalize(stmt);
+                success = true;
+            }
+
+            return success;
+        }
+
+        bool PersistentStore::getStorageSize(std::map<string, uint64_t>& namespaceSizes)
+        {
+            LOGINFO();
+
+            bool success = false;
+
+            sqlite3* &db = SQLITE;
+
+            namespaceSizes.clear();
+
+            if (db)
+            {
+                sqlite3_stmt *stmt;
+                sqlite3_prepare_v2(db, "SELECT name, sum(length(key)+length(value))"
+                                       " FROM item"
+                                       " INNER JOIN namespace ON namespace.id = item.ns"
+                                       " GROUP BY name"
+                                       ";", -1, &stmt, NULL);
+
+                while (sqlite3_step(stmt) == SQLITE_ROW)
+                    namespaceSizes[(const char*)sqlite3_column_text(stmt, 0)] = sqlite3_column_int(stmt, 1);
+
+                sqlite3_finalize(stmt);
+                success = true;
+            }
+
+            return success;
+        }
+
+        void PersistentStore::term()
+        {
+            LOGINFO();
+
+            sqlite3* &db = SQLITE;
+
+            if (db)
+                sqlite3_close(db);
+
+            db = NULL;
+        }
+
+        void PersistentStore::vacuum()
+        {
+            LOGINFO();
+
+            sqlite3* &db = SQLITE;
+
+            if (db)
+            {
+                char *errmsg;
+                int rc = sqlite3_exec(db, "VACUUM", 0, 0, &errmsg);
+                if (rc != SQLITE_OK || errmsg)
+                {
+                    if (errmsg)
+                    {
+                        LOGERR("%s", errmsg);
+                        sqlite3_free(errmsg);
+                    }
+                    else
+                        LOGERR("%d", rc);
+                }
+            }
+        }
+
+        bool PersistentStore::init(const char* filename, const char* key)
+        {
+            LOGINFO();
+
+            sqlite3* &db = SQLITE;
+
+            term();
+
+            bool shouldEncrypt = key && *key;
+#if defined(SQLITE_HAS_CODEC)
+            bool shouldReKey = shouldEncrypt && Utils::fileExists(filename) && !fileEncrypted(filename);
+#endif
+            int rc = sqlite3_open(filename, &db);
+            if (rc)
+            {
+                LOGERR("%d : %s", rc, sqlite3_errmsg(db));
+                term();
+                return false;
+            }
+
+            if (shouldEncrypt)
+            {
+#if defined(SQLITE_HAS_CODEC)
+                std::vector<uint8_t> pKey;
```

Comment:
Hi Nikita,
The code is being matched by blackduck and must be used with credit.   Please add a comment before this line as follows:
/* Based on pxCore, Copyright 2015-2018 John Robinson */
/* Licensed under the Apache License, Version 2.0 */

and append the same two lines to the end of NOTICE (in the top level).

Thanks,
Martin
Martin Hughes, RDK CMF Compliance Team